src/speculators/models/eagle.py

```diff
@@ -112,6 +169,351 @@ def validate_transformer_layer_config(cls, value: Any) -> PretrainedConfig:
             return value
 
         raise ValueError(
-            "transformer_layer_config must be a PretrainedConfig or a dict "
-            "that can be converted to a PretrainedConfig."
+            "transformer_layer_config must be a PretrainedConfig instance or a "
+            "dictionary that can be converted to a PretrainedConfig."
+        )
+
+
+@SpeculatorModel.register("eagle")
+class EagleSpeculator(SpeculatorModel):
+    """
+    A SpeculatorModel implementation for EAGLE and HASS variants for spec decoding:
+        - Eagle / Eagle v1: https://arxiv.org/abs/2401.15077
+        - Eagle v2: https://arxiv.org/abs/2406.16858
+        - HASS: https://arxiv.org/abs/2408.15766
+
+    Architecture Overview:
+        The EAGLE speculator consists of:
+        1. Input embedding layer (shared with verifier)
+        2. Optional embedding layer normalization
+        3. Fusion layer: Concatenates and projects input embeddings + verifier hidden
+           states to a latent space of hidden_size
+        4. Single transformer decoder layer for candidate token generation
+        5. Optional pre-LM head layer normalization
+        6. Language model head (shared with verifier)
+
+    Speculative Decoding Process:
+        1. Verifier model processes input and generates hidden states
+        2. EAGLE speculator uses these hidden states + input embeddings to predict
+           next tokens
+        3. Multiple candidate tokens generated in parallel using token proposal methods
+        4. Verifier validates candidates and accepts/rejects based on probability
+           thresholds
+        5. Process continues iteratively for multi-token speculation
+
+    Example:
+        ```python
+        from speculators import SpeculatorsConfig, VerifierConfig
+        from speculators.models import EagleSpeculator, EagleSpeculatorConfig
+        from speculators.proposals import GreedyTokenProposalConfig
+        from transformers import AutoConfig, AutoTokenizer
+
+        config = EagleSpeculatorConfig(
+            transformer_layer_config=AutoConfig.from_pretrained("meta-llama/Llama-3.1-8B-Instruct"),
+            speculators_config=SpeculatorsConfig(
+                algorithm="eagle_v1",
+                proposal_methods=[
+                    GreedyTokenProposalConfig(),
+                ],
+                default_proposal_method="greedy",
+                verifier=VerifierConfig(
+                    name_or_path="meta-llama/Llama-3.1-8B-Instruct",
+                    architectures=["LlamaForCausalLM"],
+                )
         )
+        speculator = EagleSpeculator(
+            config, verifier=verifier, verifier_attachment_mode="full"
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.1-8B-Instruct")
+        input_ids = tokenizer.encode("The future of AI is", return_tensors="pt")
+        outputs = speculator.generate(
```

Comment:
I think the example is a bit confusing. Is it supposed to show the initialization of a random speculator for training or of a pretrained speculator ready for inference?

It looks to me it's the latter, but in this case shouldn't we use `from_pretrained`?

Reply:
We don't have any models pushed up yet that would work with the pathways for this example. Once we have one up, we can swap for from_pretrained with a workable stub set. For now, though, I'll remove the tokenizer and generate execution